File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.java
import java.io.*;
import java.util.Arrays;
import java.util.ArrayList;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(
            new InputStreamReader(
                System.in
            )
        );

        int n = Integer.parseInt(reader.readLine().trim());
        int[] nums = Arrays.stream(
            reader.readLine().trim().split(" ")
        ).mapToInt(
            (s) -> { return Integer.parseInt(s); }
        ).toArray();
        reader.close();

        ArrayList<Integer> list = new ArrayList<>(n);
        for (int v : nums) {
            insertVal(list, v);
        }

        System.out.println(list.size());
    }

    static void insertVal(ArrayList<Integer> list, int val) {
        int b = 0;
        int e = list.size() - 1;

        int p = binarySearch(list, val, b, e);
        if (p == list.size()) {
            list.add(val);
        } else {
            list.set(p, val);
        }
    }

    static int binarySearch(ArrayList<Integer> list, int val, int b, int e) {
        if (b > e) {
            return b;
        }

        int m = (b + e) / 2;
        int pivot = list.get(m);

        if (val < pivot) {
            return binarySearch(list, val, b, m-1);
        } else if (val > pivot) {
            return binarySearch(list, val, m+1, e);
        } else {
            return m;
        }
    }
}

File: 백준/Gold/12015. 가장 긴 증가하는 부분 수열 2/가장 긴 증가하는 부분 수열 2.rs
use std::io::stdin;

fn main() {
    let mut tmp = String::new();
    stdin().read_line(&mut tmp).unwrap();
    let n: isize = tmp.trim().parse().unwrap();
    tmp.clear();

    stdin().read_line(&mut tmp).unwrap();
    let nums = tmp
        .split_whitespace()
        .into_iter()
        .map(|s| s.parse::<isize>().unwrap())
        .collect::<Vec<isize>>();

    let mut list = Vec::<isize>::new();
    
    nums.into_iter().for_each(
        |v| insert_val(&mut list, v)
    );

    println!("{:?}", list.len());
}

fn insert_val(list: &mut Vec<isize>, val: isize) {
    let b = 0;
    let e = list.len() as isize - 1;

    let p = binary_search(list, val, b, e);

    match list.get(p) {
        Some(_) => list[p] = val,
        None => list.push(val)
    }
}

fn binary_search(list: &Vec<isize>, val: isize, b: isize, e: isize) -> usize {
    if b > e { return b as usize; }

    let m = (b + e) / 2;
    let pivot = * list.get(m as usize).unwrap();

    if val < pivot {
        binary_search(list, val, b, m-1)
    } else if val > pivot {
        binary_search(list, val, m+1, e)
    } else {
        m as usize
    }
}